focused_proof.ml

```diff
@@ -452,16 +433,39 @@ let rec unfocus_proof = function
 let sequent_to_focused_sequent sequent =
     Async (Set_formula.empty, [], sequent)
 
-let proof_from_focused_proof focused_proof =
+let proof_from_focused_proof focused_proof notations sequent =
     let proof = unfocus_proof focused_proof in
-    remove_loop proof
-
-exception NonProvableSequent
-exception NonAutoProvableSequent
-
-let prove_sequent sequent =
-    let focused_sequent = sequent_to_focused_sequent sequent in
-    match prove_focused_sequent focused_sequent with
-    | Some true, Some focused_proof -> proof_from_focused_proof focused_proof
-    | Some false, _ -> raise NonProvableSequent
-    | _ -> raise NonAutoProvableSequent
+    let simplified_proof = remove_loop proof in
+    Proof.from_fully_replaced_proof notations sequent simplified_proof
+
+let rec iterate_on_notations_list sequent exponential_bound ttl = function
+    | [] -> None
+    | notations :: tail ->
+        let fully_replaced_sequent = replace_all_notations_in_sequent sequent notations in
+        let focused_sequent = sequent_to_focused_sequent fully_replaced_sequent in
+        match prove focused_sequent [] exponential_bound ttl with
+           | Some focused_proof -> Some (proof_from_focused_proof focused_proof notations sequent)
+           | None -> iterate_on_notations_list sequent exponential_bound ttl tail
+
+let rec prove_with_increasing_bound sequent cyclic_notations acyclic_notations cyclic_notations_possibilities exponential_bound ttl =
+    let notations_possibilities = List.map (fun l -> acyclic_notations @ l) cyclic_notations_possibilities in
+    has_reached_exponential_bound := false;
+    match iterate_on_notations_list sequent exponential_bound ttl notations_possibilities with
+        | None -> if !has_reached_exponential_bound || List.length cyclic_notations > 0
+            then let new_cyclic_notations_possibilities = cyclic_notations_possibilities
+                @ List.concat (List.map (fun n -> List.map (fun l -> n :: l) cyclic_notations_possibilities) cyclic_notations) in
+                prove_with_increasing_bound sequent cyclic_notations acyclic_notations new_cyclic_notations_possibilities (exponential_bound + 1) ttl
+            else (None, false)
+        | Some proof -> (Some proof, true)
+
+(* [prove_sequent sequent_with_notations] attempts to prove [sequent_with_notations.sequent]
+   and returns the result [(proof opt, is_provable)].
+   If proof option is None, then:
+   * is_provable is false if auto_prover performed an exhaustive research
+   * is_provable is true if max execution time was reached. *)
+let prove_sequent sequent_with_notations =
+    let cyclic_notations, acyclic_notations = split_cyclic_acyclic sequent_with_notations in
+    let max_execution_time_in_seconds = 3. in
+    let ttl = Sys.time () +. max_execution_time_in_seconds in
+    try prove_with_increasing_bound sequent_with_notations.sequent cyclic_notations acyclic_notations [[]] 0 ttl
+    with Ttl_exceeded | Stack_overflow -> (None, true)
```

Comment:
Stack_overflow happens on very simple example like this one https://linearon.modusponens.dev/?s=A&n=A%2CB&n=B%2CA

Reply:
I do not thing there is any point in unfolding definitions more than, let say, 10 times. Maybe adding such a bound would be nicer than waiting for `Stack_overflow` to be raised.

Reply:
You're right Stack_overflow is ugly. I changed the way I process in last commit.
If sequents after notation replace don't change (and we don't reach exponential_bound) then we can assert that the research failed after an exhaustive research and therefore mark the sequent with a red turnstile.